pkg/clusteragent/autoscaling/externalmetrics/metrics_retriever.go

```diff
@@ -160,33 +167,38 @@ func (mr *MetricsRetriever) retrieveMetricsValuesSlice(datadogMetrics []model.Da
 					datadogMetricFromStore.Error = nil
 				} else {
 					datadogMetricFromStore.Valid = false
-					datadogMetricFromStore.Error = fmt.Errorf(invalidMetricOutdatedErrorMessage, query)
+					datadogMetricFromStore.Error = NewInvalidMetricOutdatedError(query)
 				}
 			} else {
 				datadogMetricFromStore.Valid = false
 				if mr.splitBatchBackoffOnErrors {
 					incrementRetries(datadogMetricFromStore)
-					datadogMetricFromStore.Error = fmt.Errorf(invalidMetricErrorWithRetriesMessage,
-						queryResult.Error, query, datadogMetricFromStore.RetryAfter.Format(time.RFC3339))
+					datadogMetricFromStore.Error = NewInvalidMetricErrorWithRetries(queryResult.Error, query, datadogMetricFromStore.RetryAfter.Format(time.RFC3339))
 				} else {
-					datadogMetricFromStore.Error = fmt.Errorf(invalidMetricErrorMessage, queryResult.Error, query)
+					datadogMetricFromStore.Error = NewInvalidMetricError(queryResult.Error, query)
 				}
 			}
 		} else {
 			datadogMetricFromStore.Valid = false
 			if globalError {
 				if mr.splitBatchBackoffOnErrors {
 					incrementRetries(datadogMetricFromStore)
-					datadogMetricFromStore.Error = fmt.Errorf(invalidMetricGlobalErrorWithRetriesMessage, len(datadogMetrics), datadogMetricFromStore.RetryAfter.Format(time.RFC3339))
+					datadogMetricFromStore.Error = NewInvalidMetricGlobalErrorWithRetries(len(datadogMetrics), datadogMetricFromStore.RetryAfter.Format(time.RFC3339))
 				} else {
-					datadogMetricFromStore.Error = fmt.Errorf(invalidMetricGlobalErrorMessage)
+					datadogMetricFromStore.Error = NewInvalidMetricGlobalError()
 				}
 			} else {
 				// This should never happen as `QueryExternalMetric` is filling all missing series
 				// if no global error.
-				datadogMetricFromStore.Error = log.Errorf(invalidMetricNotFoundErrorMessage, query)
+				datadogMetricFromStore.Error = NewInvalidMetricNotFoundError(query)
+				log.Error(datadogMetricFromStore.Error)
 			}
 		}
+		// If there is a rate limit error, apply the rate limit error to all datadogMetrics
+		if rateLimitErr != nil {
```

Comment:
Not sure this is the right location. You'll still go through previous `else {}` that invalidates the DDM (which it should not). I believe you need some dedicated code either in an `else if {}` or in the large `else {}`